cpp/src/reshape/one_hot_encoding.cu

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/transform.h>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda/std/cmath>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+template <typename InputType>
+struct one_hot_encoding_functor {
+  one_hot_encoding_functor(column_device_view input, column_device_view category)
+    : d_input(input), d_category(category)
+  {
+  }
+
+  bool __device__ operator()(size_type i)
+  {
+    auto element_index  = i % d_input.size();
+    auto category_index = i / d_input.size();
+
+    if (d_category.is_valid(category_index) and d_input.is_valid(element_index)) {
+      if constexpr (is_floating_point<InputType>()) {
+        if (cuda::std::isnan(d_category.element<InputType>(category_index)) and
+            cuda::std::isnan(d_input.element<InputType>(element_index)))
```

Comment:
Hmmm, we've typically made a point to not treat NaNs as equal in libcudf

Reply:
Thanks for pointing this out. Let me try preprocess `categories` argument to respect that consistency.

Reply:
Well `element_equality_comparator` _already_ treats NaN as equal to NaN. So no need for preprocessing, just use the `element_equality_comparator`.